lib/Transforms/Instrumentation/ExtendedPointer.cc

```diff
@@ -201,41 +206,70 @@ bool ExtendedPointer::instrument(Value *Ptr, Value *InstVal,
 /* Record information of address Ptr, store/update the base address and size */
 bool ExtendedPointer::instrumentAddress(Value *Ptr, const DataLayout &DL,
                                         Value *BasePtr) {
+  Ptr->setName("raw_ptr");
+  Type *resultType;
   if (!BasePtr) {
-    Type *IntTy = DL.getIntPtrType(Ptr->getType());
-    // set_fat_slot0(Ptr, Base)
-    Builder->CreateCall(
+    if (auto *ALI = dyn_cast<AllocaInst>(Ptr)) {
+       resultType = ALI->getAllocatedType();
+    } else {
+      llvm_unreachable("only handling GEP instructions");
+    }
+    CallInst* withBase = Builder->CreateCall(
       m_setFatSlot0,
       {
-        Builder->CreateBitCast(Ptr, Builder->getInt8PtrTy()),
-        Builder->CreatePtrToInt(Ptr, Builder->getInt32Ty())
+        Constant::getNullValue(Builder->getInt8PtrTy()),
+        ConstantInt::get(IntptrTy, 0)
+      }
+    );
+    // set_fat_slot1(Ptr, Size)
+    CallInst* withSize = Builder->CreateCall(
+      m_setFatSlot1,
+      {
+        Constant::getNullValue(Builder->getInt8PtrTy()),
+        ConstantInt::get(IntptrTy, 0)
       }
     );
+    Value* casted = Builder->CreateBitCast(withSize, resultType->getPointerTo());
+    Ptr->replaceAllUsesWith(casted);
     Value *sizeVal;
     // alloca
     if (auto *ALI = dyn_cast<AllocaInst>(Ptr)) {
       uint64_t size = DL.getTypeStoreSize(ALI->getAllocatedType());
-      sizeVal = ConstantInt::get(Type::getInt32Ty(*Ctx), size);
+      sizeVal = ConstantInt::get(IntptrTy, size);
     } else {
       llvm_unreachable("unexpected lack of base address");
     }
-    // set_fat_slot1(Ptr, Size)
-    Builder->CreateCall(
-      m_setFatSlot1,
-      {
-        Builder->CreateBitCast(Ptr, Builder->getInt8PtrTy()),
-        sizeVal
-      }
-    );
+    // Type *IntTy = DL.getIntPtrType(Ptr->getType());
+    // set_fat_slot0(Ptr, Base)
+    Builder->SetInsertPoint(withBase);
+    Value *argA = Builder->CreateBitCast(Ptr, Builder->getInt8PtrTy());
+    Value *argB = Builder->CreatePtrToInt(Ptr, IntptrTy);
+    withBase->setArgOperand(0, argA);
+    withBase->setArgOperand(1, argB);
+    Builder->SetInsertPoint(withSize);
+    argA = Builder->CreateBitCast(withBase, Builder->getInt8PtrTy());
+    withSize->setArgOperand(0, argA);
+    withSize->setArgOperand(1, sizeVal);
   } else {
     // copy_fat_slots(Ptr, BasePtr)
-    Builder->CreateCall(
+    if (auto *GEP = dyn_cast<GetElementPtrInst>(Ptr)) {
+       resultType = GEP->getResultElementType();
+    } else {
+      llvm_unreachable("only handling GEP instructions");
+    }
+    CallInst* copied = Builder->CreateCall(
```

Comment:
need comment on this as well. Looks like code is first inserted with placeholders and then updated as needed.
Don't remember why we designed it this way.

Reply:
This would copy size and base of `BasePtr` to `Ptr`. If we use fat ptr encoding implementation, then we need to replace all usage of `Ptr` with a new pointer with copied encoding. I am also kind of lost why we needed to set arguments at the end, but I don't think these steps of replacing all usage of `Ptr` with `copied` would be needed if we just use the hashmap runtime implementation. I can revert this part (`instrumentAddress`) to an older / simpler version if we plan to just use the hashmap implementation for now.